src/common/eigf.h

```diff
@@ -0,0 +1,302 @@
+ * halos problems, and gives a smooth result.
+ * Weighting by 1 / pixel_value^2 is empirical.
+ * It was found to work well, while weighting by 1/pixel_value gives sometimes
+ * worse results.
+ * The idea of using a weighted averaging here comes from:
+ * C. N. Ochotorena and Y. Yamashita, "Anisotropic Guided Filtering," in IEEE Transactions on Image Processing, vol. 29, pp. 1397-1412, 2020, doi: 10.1109/TIP.2019.2941326.
+ * although we don't use the same weighting system (their weight is based
+ * on variance, ours is based on the pixel value).
+ * We tried a mixed weight using their weight in combination to ours, but
+ * it did not improve the result.
+ *
+ * The implementation EIGF uses downscaling to speed-up the filtering,
+ * just like what is done in fast_guided_filter.h
+**/
+
+static inline void exposure_independent_guided_filter(const float *const restrict guide, // I
+                                    const float *const restrict mask, //p
+                                    float *const restrict ab,
+                                    const size_t width, const size_t height,
+                                    const float sigma, const float feathering)
+{
+  // We also use gaussian blurs instead of the square blurs of the guided filter
+  const size_t Ndim = width * height;
+  float *const restrict blurred_guide = dt_alloc_sse_ps(Ndim);
+  // guide_x_guide = (guide - blurred_guide)^2
+  float *const restrict guide_x_guide = dt_alloc_sse_ps(Ndim);
+  // guide_variance = blur(guide_x_guide)
+  float *const restrict guide_variance = dt_alloc_sse_ps(Ndim);
+  float *const restrict guide_x_mask = dt_alloc_sse_ps(Ndim);
+  // guide_mask_covariance = blur(guide_x_mask)
+  float *const restrict guide_mask_covariance = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_mask = dt_alloc_sse_ps(Ndim);
+  float *const restrict a = dt_alloc_sse_ps(Ndim);
+  float *const restrict b = dt_alloc_sse_ps(Ndim);
+  // weight to compute the weighted blur of a and b
+  float *const restrict weights = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_a = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_b = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_weights = dt_alloc_sse_ps(Ndim);
+
+  float ming = 10000000.0f;
+  float maxg = 0.0f;
+  float minm = 10000000.0f;
+  float maxm = 0.0f;
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+dt_omp_firstprivate(guide, mask, Ndim) \
+  schedule(simd:static) aligned(guide, mask:64) \
+  reduction(max:maxg, maxm)\
+  reduction(min:ming, minm)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    const float pixelg = guide[k];
+    const float pixelm = mask[k];
+    if(pixelg < ming) ming = pixelg;
+    if(pixelg > maxg) maxg = pixelg;
+    if(pixelm < minm) minm = pixelm;
+    if(pixelm > maxm) maxm = pixelm;
+  }
+
+  dt_gaussian_t *g = dt_gaussian_init(width, height, 1, &maxg, &ming, sigma, 0);
```

Comment:
So you blur the guide and the image first, doesn't that create halos around edges ?

Reply:
The first blurs are here to compute the esperance, which is in particular used to compute variance (you have these blurs also in the original guided filter, that's your tmp[0] and tmp[1] in your implementation)